File: 307-range-sum-query-mutable/307-range-sum-query-mutable.cpp
class NumArray {
public:
    vector<int> arr;
    vector<int> memo;
    int len;
    
    NumArray(vector<int>& nums) {
        arr = nums;
        len = ceil(nums.size() / sqrt(nums.size()));
        memo = vector<int> (len, 0);
        for (int i = 0; i < nums.size(); ++i) {
            memo[i / len] += nums[i];
        }
    }
    
    void update(int index, int val) {
        int idx = index / len;
        memo[idx] = memo[idx] - arr[index] + val;
        arr[index] = val;
    }
    
    int sumRange(int left, int right) {
        int sum = 0;
        int start = left / len;
        int end = right / len;
        if (start == end) {
            for (int i = left; i <= right; ++i) {
                sum += arr[i];
            }
        }
        else {
            int left_bound = (start + 1) * len;
            int right_bound = end * len;
            
            for (int i = left; i < left_bound; ++i) {
                sum += arr[i];
            }
            for (int i = start + 1; i < end; ++i) {
                sum += memo[i];
            }
            for (int i = right_bound; i <= right; ++i) {
                sum += arr[i];
            }
        }
        return sum;
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */

File: 307-range-sum-query-mutable/307-range-sum-query-mutable.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = nums
        self.len = int(math.ceil(len(nums) / math.sqrt(len(nums))))
        self.memo = [0] * self.len
        for i in range(len(nums)):
            self.memo[int(i / self.len)] += nums[i]
        
    def update(self, index: int, val: int) -> None:
        idx = int(index / self.len)
        self.memo[idx] = self.memo[idx] - self.nums[index] + val
        self.nums[index] = val

    def sumRange(self, left: int, right: int) -> int:
        sum_ = 0
        start = int(left / self.len)
        end = int(right / self.len)
        if start == end:
            for i in range(left, right + 1):
                sum_ += self.nums[i]
        else:
            left_bound = (start + 1) * self.len
            right_bound = end * self.len
            
            for i in range(left, left_bound):
                sum_ += self.nums[i]
            for i in range(start + 1, end):
                sum_ += self.memo[i]
            for i in range(right_bound, right + 1):
                sum_ += self.nums[i]
        return sum_

# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)